feat(#560): edit coordinates for a vertex or feature

latin-panda opened the pull request:
Closes #560

### I have verified this PR works in these browsers (latest versions):

- [x] Chrome
- [x] Firefox
- [x] Safari (macOS)
- [ ] Safari (iOS)
- [x] Chrome for Android
- [ ] Not applicable

### What else has been done to verify that this works as intended?

<details><summary>Test case: Errors messages in desktop</summary>

<img width="800" src="https://github.com/user-attachments/assets/c1f73a1b-5b86-46f7-984a-bafddb8d9b13" />
<img width="800" src="https://github.com/user-attachments/assets/823a30a2-fe48-4f10-a96b-e4d7dbea3e2d" />
<img width="800" src="https://github.com/user-attachments/assets/6971c688-94f9-4f44-8cff-9b449974eee5" />
<img width="800" src="https://github.com/user-attachments/assets/f14373a1-e726-47eb-bca8-c34fd8bddb71" />

</details>

<details><summary>Test case: Errors messages in mobile</summary>

<img width="400" src="https://github.com/user-attachments/assets/c353364b-7f7c-475a-bb66-0dd784daea22" />
<img width="400" src="https://github.com/user-attachments/assets/3bc95a9e-a065-4f5f-837f-5f21dc4e962c" />
<img width="400" src="https://github.com/user-attachments/assets/7ea1a021-24f4-46b1-befa-45769daa3fc5" />
<img width="400" src="https://github.com/user-attachments/assets/287f2611-8f3c-48df-aa0d-27c3e82f9e83" />

</details>

<details><summary>Test case: Full-screen experience</summary>

<img width="800" src="https://github.com/user-attachments/assets/afd14331-6b62-4d09-bb85-e72d516a4168" />
<img width="800" src="https://github.com/user-attachments/assets/e0c70ccb-8845-43a6-a1ed-5403fda92729" />

<img width="400" src="https://github.com/user-attachments/assets/14cd1b49-961d-4bb6-ad37-bc9605b92d57" />
<img width="400" src="https://github.com/user-attachments/assets/9103ded4-39ad-49b6-b8c9-6e2ce51cea8a" />

</details>

<details><summary>Test case: readonly map</summary>

https://github.com/user-attachments/assets/ac2d51f5-b606-4165-9ad9-297bb6ec88e6

</details>

<details><summary>Test case: geoshape</summary>

https://github.com/user-attachments/assets/9f95d9e2-7287-444a-a3c5-83f5cb1f0576

</details>

<details><summary>Test case: geotrace</summary>

https://github.com/user-attachments/assets/ade76db4-b4f8-4a63-906c-7497e143ba39

</details>
 
<details><summary>Regression test: geopoint with "maps" and "placement-map" appearances</summary>

https://github.com/user-attachments/assets/a83764bf-ccbf-4eec-9858-dfe853687e34

</details>

<details><summary>Regression test: select with "map" appearance</summary>

https://github.com/user-attachments/assets/b42b8518-6b4b-448d-9470-17f1e0754640

</details>

<details><summary>Central smoke test</summary>

https://github.com/user-attachments/assets/912a2399-3fb8-4023-8d55-c7ad3c1e7a28

</details>

<details><summary>Bundle size verification</summary>

No big difference means OL didn't leak, because OL would inflate the main chunk by 400 or 600.

This branch:

<img width="900" src="https://github.com/user-attachments/assets/087f7706-e7c5-4bb4-8492-406128bf1ab8" />

In `main` branch:

<img width="900" src="https://github.com/user-attachments/assets/1836f83d-b613-4c5f-85bb-86eeeb83c2ab" />

</details>

### Why is this the best possible solution? Were any other approaches considered?

This is an extension of the existent solution:
- Follows the same pattern of small components for modularity and ease of maintenance 
- Follows the same pattern of adding new features as configurable capabilities in each mode type. 
- Reuses existing utilities without having side effects on existing features due to small pure functions with defined responsibilities. 

### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

- Mostly extension of code, regression test is done in case an unintended reactive value state is in the selected feature/vertex 

### Do we need any specific form for testing your changes? If so, please attach one.

- It can be tested using existing test forms. 

### What's changed
- Adds a dialog component to allow users to paste a value or upload a GeoJSON file
- Adds an expandable component to display advanced features under the map
- Updates the status bar component to fit a new button that displays the advanced features panel  
- Adds a new utility function in `mapHelpers.ts` to perform basic validation and return coordinates based on the expected feature type. This is used to return the correct coordinate format after the user pastes a value or uploads a file.
- Exposes existing utility functions that parse GeoJSON format for reusability and renames file from `createFeatureCollectionAndProps.ts` to ‎`geojson-parsers.ts`
- Adds utility functions to parse the first feature in a GeoJSON file to `geojson-parsers.ts`
- Adds two new capabilities to the mode config of the map: 'canUpdateFeatureCoordinates' and 'canUpdateVertexCoordinates'
- Adds test coverage and updates visual e2e tests

latin-panda commented on packages/web-forms/src/components/common/map/useMapBlock.ts:
```diff
@@ -206,6 +211,9 @@ export function useMapBlock(config: MapBlockConfig, events: MapBlockEvents) {
 		mapFeatures?.saveFeature(feature);
+		// Refresh selected vertex.
+		const vertexIndex = feature.get(SELECTED_VERTEX_INDEX_PROPERTY) as number | undefined;
+		events.onVertexSelect(getVertexByIndex(feature, vertexIndex));
```
This refresh updates the status bar and the advanced panel after a vertex is deleted or updated.

latin-panda commented on packages/web-forms/src/components/common/map/getModeConfig.ts:
```diff
@@ -6,6 +11,18 @@ export const MODES = {
 export type Mode = (typeof MODES)[keyof typeof MODES];
 
+// Used when loading a single feature from the map.
+export const SINGLE_FEATURE_TYPES = {
```
It's called "single" because we also support maps with a mix of feature types at the same time (select with the `map` appearance), but for geopoint, geoshape, and geotrace, we only support one type at a time.

sadiqkhoja reviewed (commented):
Few observations while trying the demo:

- Long press requires quite a long press, is there a way to reduce the time
- Undo button allows only 1 step back, doesn't it create editing stack?
- The +/- button in Longitude and Latitude textbox are not super useful, would be best to remove them.
- If two vertices are exactly same then how can I select one or other? Maybe next/prev button would be helpful

sadiqkhoja commented on packages/web-forms/e2e/test-cases/visual/all-question-types.test.ts-snapshots/geoshape-autoclose-polygon-chromium.png:
Meta-comment: maybe we should configure LFS for these images.

sadiqkhoja commented on packages/web-forms/src/components/common/map/getModeConfig.ts:
```diff
@@ -6,6 +11,18 @@ export const MODES = {
+	TRACE: 'trace',
+} as const;
+export type SingleFeatureType = (typeof SINGLE_FEATURE_TYPES)[keyof typeof SINGLE_FEATURE_TYPES];
+export const VERTEX_FEATURES: SingleFeatureType[] = [
```
nit: maybe `GEO_OBJECT_TYPES` or `GEOMETRY_TYPES`? Shape/trace is not really feature of a vertex.

latin-panda commented:
@alyblenkin Thanks for the feedback! 

> Can we tighten the spacing between the text field and error so they look more connected?

Sure, fixed, same as Figma
<details><summary>Expand to see fix</summary>

<img width="800"  src="https://github.com/user-attachments/assets/c6e96dfa-eb2f-4b54-92eb-0593368c8c75" />

</details>

> When importing a file, can we also drag and drop? Is that harder to implement right now? Not a deal breaker!

No right now, we don't even have it in the upload image question type. It would be good to implement it together. I've created a [ticket](https://github.com/getodk/web-forms/issues/622).

> I realized we should probably link to the docs to explain the ODK format. Maybe this link.

It's better to avoid adding links, since they can break easily or change paths, and fixing them will require a release in Web Forms and Central. For now, I've added an icon with a tooltip that shows this information on hover as a quick solution.

<details><summary>Expand to see fix</summary>

<img width="800" src="https://github.com/user-attachments/assets/388c463a-2b70-45a2-b24c-795ac591a193" />

</details>

> Expand text field for pasting location data so you can view the entire string

This is fixed

<details><summary>Expand to see fix</summary>

<img width="800" src="https://github.com/user-attachments/assets/7884ddf5-9bd4-43d9-b714-5c97fd000b62" />

<img width="800" src="https://github.com/user-attachments/assets/86093d59-6a82-4776-9628-fbff610118ea" />

</details>

latin-panda commented:
@sadiqkhoja I've addressed all the feedback, and this is ready for another round of review ;)

alyblenkin commented:
> @alyblenkin I've added the file name

Oh thank you! That's awesome. I think we don't need to say "file added" but the f[ile name is helpful. ](https://www.figma.com/design/6v0NnvayfDlaQhYpjv37Bn/ODK-Web-Forms?node-id=4174-20184&t=B9qJYyiWW4nC6nmh-1)
